contracts/adapters/CharmAdapter.sol

```diff
@@ -0,0 +1,388 @@
+     */
+    function exercise(
+        address options,
+        uint256 optionID,
+        uint256 amount,
+        address recipient
+    ) public payable override {
+
+      // Get strike index, whether is long
+      OptionType memory optionType = addressToOptionType[options];
+
+      //get token
+      IOptionToken token = IOptionToken(options);
+      //get market
+      IOptionMarket market = IOptionMarket(token.market());
+
+      uint256 profit = exerciseProfit(options, optionID, amount);
+      require(profit > 0, "Not profitable to exercise");
+
+      // if we are exercising but market has not settled, do it
+      if(market.isExpired() && !market.isSettled()){
+        market.settle();
+      }
+
+      uint256 amountOut = market.sell(optionType.isLongToken, optionType.strikeIndex, amount, profit);
+
+      //transfer over
+      market.baseToken().uniTransfer(payable(recipient), amountOut);
+
+      emit Exercised(
+          msg.sender,
+          options,
+          optionID,
+          amount,
+          amountOut
+      );
+    }
+
+    function createShort(ProtocolAdapterTypes.OptionTerms memory, uint256)
+        external
+        pure
+        override
+        returns (uint256)
+    {
+        return 0;
+    }
+
+    function closeShort() external pure override returns (uint256) {
+        return 0;
+    }
+
+    /**
+     * @notice Function to lookup oToken addresses.
+     * @param optionTerms is the terms of the option contract
+     */
+    function lookupOToken(ProtocolAdapterTypes.OptionTerms memory optionTerms)
+        public
+        view
+        returns (address token)
+    {
+        bool isPut =
+            optionTerms.optionType == ProtocolAdapterTypes.OptionType.Put;
+
+        //market.baseToken() is underlying asset if call. Strike currency if put. Represents ETH if equal to 0x0
+        address underlying = isPut == true ? optionTerms.strikeAsset : optionTerms.underlying;
+
+        //there should only be a collateral asset for charm if we are writing an option
+        bool isShort = optionTerms.collateralAsset == address(1) ? false : true;
+
+        // refer to mapping after encoding with _getOptionId
+        bytes32 id = _getOptionId(underlying, isShort, optionTerms.strikePrice, optionTerms.expiry, isPut);
+
+        token = idToAddress[id];
+    }
+
+    // Populate mappings of option type to option token address
+    function populateOTokenMappings(ProtocolAdapterTypes.OptionTerms memory optionTerms)
+        internal
+    {
+
+      bool isPut =
+          optionTerms.optionType == ProtocolAdapterTypes.OptionType.Put;
+
+      //market.baseToken() is underlying asset if call. Strike currency if put. Represents ETH if equal to 0x0
+      address underlying = isPut == true ? optionTerms.strikeAsset : optionTerms.underlying;
+
+      uint256 i = 0;
+
+      address[] memory markets = optionFactory.markets();
+
+      while (i < markets.length) {
```

Comment:
One idea to optimize this is to iterate in reverse, and then when we hit an OptionMarket that was already stored, break from the loop. That saves us a lot of gas from iterating from index 0, which is already seen when called a second time.

```solidity
i = markets.length - 1

while (i != 0) {
 if (seen) break

 if (not seen) {
  populate()
  continue
 }

  i -= 1
}
```

Reply:
thats clever